src/ddscxx/include/dds/sub/detail/SampleRef.hpp

```diff
@@ -0,0 +1,117 @@
+#ifndef OMG_DDS_PUB_DETAIL_SAMPLEREF_HPP_
+#define OMG_DDS_PUB_DETAIL_SAMPLEREF_HPP_
+
+/* Copyright 2010, Object Management Group, Inc.
+ * Copyright 2010, PrismTech, Corp.
+ * Copyright 2010, Real-Time Innovations, Inc.
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+namespace dds
+{
+namespace sub
+{
+namespace detail
+{
+template <typename T>
+class SampleRef;
+}}}
+
+#include <dds/sub/SampleInfo.hpp>
+#include <org/eclipse/cyclonedds/topic/datatopic.hpp>
+
+namespace dds
+{
+namespace sub
+{
+namespace detail
+{
+template <typename T>
+class SampleRef
+{
+public:
+    SampleRef()
+    {
+      this->data_ = nullptr;
+    }
+
+    SampleRef(ddscxx_serdata<T>* d, const dds::sub::SampleInfo& i)
+    {
+        this->data_ = d;
+        this->info_ = i;
+    }
+
+    SampleRef(const SampleRef& other)
+    {
+        copy(other);
+    }
+
+    virtual ~SampleRef()
+    {
+        ddsi_serdata_unref(data_);
+    }
+
+    SampleRef& operator=(const SampleRef& other)
+    {
+        return copy(other);
+    }
+
+    SampleRef& copy(const SampleRef& other)
+    {
+        this->data_ = other.data_;
+        this->info_ = other.info_;
+
+        return *this;
+    }
+
+public:
+    const T& data() const
+    {
+        return *data_->getT();
+    }
+
+    const dds::sub::SampleInfo& info() const
+    {
+        return info_;
+    }
+
+    dds::sub::SampleInfo& info()
+    {
+        return info_;
+    }
+
+    bool operator ==(const SampleRef& other) const
```

Comment:
Good catch. This is actually a by-product from copy-pasting most of the SampleRef from Sample, where Sample also didn't implement the == operator. For the latter there is actually an internal ticket (OSPL-2430) describing the reason for not implementing it: namely the fact that the old IDL compiler didn't generate == operators for the generated sample datatypes yet, and since the == operator of Sample will need to delegate to the == operator of both the SampleInfo and the sample it couldn't be done. With the advent of the new IDL compiler this is no longer an issue, and should be solved as well. For SampleRef I will do that right here, for Sample I guess we should do that as a separate issue.

Reply:
It turns out the current IDL compiler still doesn't implement the == operator on the data samples. Therefore I propose to fix this issue in a separate pull request that not only covers the generation of == operators on the generated data samples, but also the correct implementation of the == operators for dds::sub::Sample and for dds::sub::SampleRef.

Reply:
Proposal accepted 🙂